---
title: >
  Using GeneTonic on the macrophage dataset (EGAS00001000563)
author:
- name: Federico Marini
  affiliation: 
  - &id1 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz<br>
  - &id2 Center for Thrombosis and Hemostasis (CTH), Mainz<br>
  email: marinif@uni-mainz.de
- name: Annekathrin Ludt
  affiliation: 
  - *id1
- name: Jan Linke
  affiliation: 
  - *id1
  - *id2
- name: Konstantin Strauch
  affiliation: 
  - *id1

date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('ideal')`"
output: 
  bookdown::html_document2:
    toc: true
    toc_float: true
    theme: cosmo
    code_folding: show
    code_download: true
editor_options: 
  chunk_output_type: console
link-citations: true
bibliography: "../genetonic_supplement.bib"
---

```{r setup, include=FALSE, cache=FALSE, eval = TRUE, echo = FALSE}
library("knitr")
opts_chunk$set(
  fig.align = "center",
  fig.show = "asis",
  eval = TRUE,
  fig.width = 10,
  fig.height = 7,
  tidy = FALSE,
  message = FALSE,
  warning = FALSE,
  size = "small",
  comment = "##",
  echo = TRUE,
  results = "markup"
)
options(replace.assign = TRUE, width = 80)
```

# About the data

The data illustrated in this document is an RNA-seq dataset, available at the European Genome-Phenome archive under the accession code EGAS00001002236 (https://ega-archive.org/studies/EGAS00001002236).

The data is included in the `macrophage` package available on Bioconductor (https://www.bioconductor.org/packages/release/data/experiment/html/macrophage.html). The data was generated as part of the work to identify shared quantitative trait loci (QTLs) for chromatin accessibility and gene expression in human macrophages [@Alasoo2018] - the manuscript is available at https://www.nature.com/articles/s41588-018-0046-7.


# Loading required packages

We load the packages required to perform all the analytic steps presented in this document.

```{r loadLibraries, results='hide'}
library("DESeq2")
library("topGO")
library("pheatmap")
library("org.Hs.eg.db")
library("pcaExplorer")
library("ideal")
library("DT")
library("GeneTonic")
library("apeglm")
library("dplyr")
library("clusterProfiler")
library("gprofiler2")
library("tibble")
library("fgsea")
library("msigdbr")
library("enrichR")
library("visNetwork")
library("magrittr")
```

# Data processing

After obtaining the data from the `macrophage` package, we generated a `DESeqDataset` object, provided in this repository. The generation of this object is demonstrated in the `GeneTonic` package vignette available through `browseVignette(GeneTonic)` after successful installation of the package. As in the vignette of `GeneTonic`, we will restrict our intention to the comparison between Interferon gamma treated samples vs naive samples of the `macrophage` package.

```{r create_dds, eval=FALSE}
library("macrophage")
library("DESeq2")

data("gse", package = "macrophage")

dds_macrophage <- DESeqDataSet(gse, design = ~line + condition)
# changing the ids to Ensembl instead of the Gencode used in the object
rownames(dds_macrophage) <- substr(rownames(dds_macrophage), 1, 15)
dds_macrophage

# saveRDS(dds_macrophage, file = "usecase_dds_macrophage.rds")
```

## Exploratory data analyisis

We read in the dataset, apply the vst transformation for performing PCA and creating a heatmap of the sample to sample distances.  
We'll use some functions from the `pcaExplorer` package [@Marini2019].

```{r eda-macrophage}
dds_macrophage <- readRDS("usecase_dds_macrophage.rds")
dds_macrophage

vst_macrophage <- vst(dds_macrophage)
vst_macrophage <- vst_macrophage[, vst_macrophage$condition %in% c("naive", "IFNg")]

pheatmap::pheatmap(as.matrix(dist(t(assay(vst_macrophage)))))

pcaExplorer::pcaplot(vst_macrophage,
                     ntop = 5000,
                     title = "PCA - top5000 most variable genes",
                     ellipse = FALSE
                     )
```

## Differential expression analysis

We set the False Discovery Rate to 0.01 and we run the `DESeq2` workflow, generating results and using the `apeglm` shrinkage estimator.  
We plot the results as an MA-plot and report them in a table, using the functions from the `ideal` package [@Marini2020].

```{r de-macrophage}
anno_df <- readRDS("usecase_annodf_macrophage.rds")
FDR <- 0.01

dds_macrophage <- DESeq(dds_macrophage)

results_IFNg_vs_naive <- results(dds_macrophage,
                                 name = "condition_IFNg_vs_naive",
                                 lfcThreshold = 1,
                                 alpha = FDR)
summary(results_IFNg_vs_naive)

results_IFNg_vs_naive$SYMBOL <- anno_df$gene_name[match(rownames(results_IFNg_vs_naive), anno_df$gene_id)]

ideal::plot_ma(results_IFNg_vs_naive, ylim = c(-5, 5), title = "MAplot - INFg vs Naive")

tbl_DEres_IFNg_vs_naive <- deseqresult2df(results_IFNg_vs_naive, FDR = FDR)

DT::datatable(tbl_DEres_IFNg_vs_naive, rownames = FALSE)
```

## Functional enrichment analysis

We perform functional enrichment analysis, here using the `topGOtable` wrapper to the method implemented in the `topGO` package.


```{r enrich-alma, cache=TRUE}
expressedInAssay <- (rowSums(assay(dds_macrophage)) > 0)
geneUniverseExprENS <- rownames(dds_macrophage)[expressedInAssay]
geneUniverseExpr <- anno_df$gene_name[match(geneUniverseExprENS, anno_df$gene_id)]

GObps_INFg_vs_naive <- topGOtable(
  DEgenes = tbl_DEres_IFNg_vs_naive$SYMBOL,
  BGgenes = geneUniverseExpr,
  ontology = "BP",
  geneID = "symbol",
  addGeneToTerms = TRUE,
  mapping = "org.Hs.eg.db",
  topTablerows = 500
)

GOmfs_INFg_vs_naive <- topGOtable(
  DEgenes = tbl_DEres_IFNg_vs_naive$SYMBOL,
  BGgenes = geneUniverseExpr,
  ontology = "MF",
  geneID = "symbol",
  addGeneToTerms = TRUE,
  mapping = "org.Hs.eg.db",
  topTablerows = 500
)

GOccs_INFg_vs_naive <- topGOtable(
  DEgenes = tbl_DEres_IFNg_vs_naive$SYMBOL,
  BGgenes = geneUniverseExpr,
  ontology = "CC",
  geneID = "symbol",
  addGeneToTerms = TRUE,
  mapping = "org.Hs.eg.db",
  topTablerows = 500
)

res_enrich_macrophage <- shake_topGOtableResult(GObps_INFg_vs_naive)
res_enrich_macrophage <- get_aggrscores(
  res_enrich = res_enrich_macrophage,
  res_de = results_IFNg_vs_naive,
  annotation_obj = anno_df
)


DT::datatable(res_enrich_macrophage, rownames = FALSE)


res_enrich <- shake_topGOtableResult(GObps_INFg_vs_naive)
res_enrich <- get_aggrscores(
  res_enrich = res_enrich,
  res_de = results_IFNg_vs_naive,
  annotation_obj = anno_df
)

saveRDS(res_enrich, file = "usecase_res_enrich_macrophage.rds")

DT::datatable(res_enrich, rownames = FALSE)
```

### Using alternative enrichment analysis methods

It is possible to use the output of different other methods for enrichment analysis, thanks to the `shaker_*` functions implemented in `GeneTonic` - see for example `?shake_topGOtableResult`, and navigate to the manual pages for the other functions listed in the "See Also" section.

`GeneTonic` is able to convert for you the output of

- DAVID (text file, downloaded from the DAVID website)
- clusterProfiler (takes `enrichResult` objects)
- enrichr (a data.frame with the output of `enrichr`, or the text file exported from Enrichr)
- fgsea (the output of `fgsea()` function)
- g:Profiler (the text file output as exported from g:Profiler, or a data.frame with the output of `gost()` in `gprofiler2`)

Some examples on how to use them are reported here:

```{r clusterProfiler}
# clusterProfiler --------------------------------------------------------------
degenes_macrophage <- deseqresult2df(results_IFNg_vs_naive, FDR = FDR)$id
ego_macrophage <- enrichGO(gene          = degenes_macrophage,
                     universe      = rownames(dds_macrophage)[expressedInAssay],
                     OrgDb         = org.Hs.eg.db,
                     ont           = "BP",
                     keyType       = 'ENSEMBL',
                     pAdjustMethod = "BH",
                     pvalueCutoff  = 0.01,
                     qvalueCutoff  = 0.05,
                     readable      = TRUE)
res_enrich_clusterprofiler <- shake_enrichResult(ego_macrophage)

# g:profiler -------------------------------------------------------------------
degenes <- deseqresult2df(results_IFNg_vs_naive, FDR = FDR)$SYMBOL
gostres_macrophage <- gost(
  query = degenes, 
  ordered_query = FALSE, 
  multi_query = FALSE, 
  significant = FALSE, 
  exclude_iea = TRUE, 
  measure_underrepresentation = FALSE, 
  evcodes = TRUE, 
  user_threshold = 0.05, 
  correction_method = "g_SCS", 
  domain_scope = "annotated", 
  numeric_ns = "", 
  sources = "GO:BP", 
  as_short_link = FALSE)

res_enrich_gprofiler <- shake_gprofilerResult(gprofiler_output = gostres_macrophage$result)

# fgsea ------------------------------------------------------------------------
res2 <- results_IFNg_vs_naive %>%
  as.data.frame() %>% 
  dplyr::select(SYMBOL, log2FoldChange)
de_ranks <- deframe(res2)
head(de_ranks, 20)
msigdbr_df <- msigdbr(species = "Homo sapiens", category = "C5", subcategory = "BP")
msigdbr_list = split(x = msigdbr_df$gene_symbol, f = msigdbr_df$gs_name)
fgseaRes <- fgsea(pathways = msigdbr_list, 
                  stats = de_ranks, 
                  nperm=100000)
fgseaRes <- fgseaRes %>% 
  arrange(desc(NES))

res_enrich_fgsea <- shake_fgseaResult(fgsea_output = fgseaRes)

# enrichr ----------------------------------------------------------------------
dbs <- c("GO_Molecular_Function_2018",
         "GO_Cellular_Component_2018",
         "GO_Biological_Process_2018",
         "KEGG_2019_Human",
         "Reactome_2016",
         "WikiPathways_2019_Human")
degenes <- (deseqresult2df(results_IFNg_vs_naive, FDR = FDR)$SYMBOL)
enrichr_output_macrophage <- enrichr(degenes, dbs)

res_enrich_enrichr_BPs <- shake_enrichrResult(
  enrichr_output = enrichr_output_macrophage$GO_Biological_Process_2018)
res_enrich_enrichr_KEGG <- shake_enrichrResult(
  enrichr_output = enrichr_output_macrophage$KEGG_2019_Human)
```

## Assembling the `gtl` object

To simplify the usage of the function from `GeneTonic`, we create a `GeneTonic_list` object.

```{r gtllist, eval = F}
gtl_macrophage <- GeneTonic_list(
  dds = dds_macrophage,
  res_de = results_IFNg_vs_naive,
  res_enrich = res_enrich,
  annotation_obj = anno_df
)

saveRDS(gtl_macrophage, "gtl_macrophage.rds")
```

# Running `GeneTonic` on the dataset

Let's load the `GeneTonicList` object, containing all the structured input needed.


```{r, eval = F}
gtl_macrophage <- readRDS("gtl_macrophage.rds")
```

## `GeneTonic`, interactively

This command will launch the `GeneTonic` app:

```{r eval=FALSE}
GeneTonic(gtl = gtl_macrophage)
```

The exploration can be guided by launching the introductory tours for each section of the app, and finally one can generate a small report focused on the features and genesets of interest.

If some more custom visualizations are required, it is possible to export the objects in use from the app as a `SummarizedExperiment` object, to be then passed to the `iSEE` software [@Rue-Albrecht2018] for further exploration.

## Using `GeneTonic`'s functions in analysis reports 

The functionality of `GeneTonic` can be used also as standalone functions, to be called for example in existing analysis reports in RMarkdown, or R scripts.  
In the following chunks, we show how it is possible to call some of the functions on the dataset of the `macrophage` package.

```{r graphs, eval = F}
em <- enrichment_map(gtl = gtl_macrophage, n_gs = 100)


em %>%
  visIgraph() %>%
  visOptions(
    highlightNearest = list(
      enabled = TRUE,
      degree = 1,
      hover = TRUE
    ),
    nodesIdSelection = TRUE
  )

GeneTonic::gs_heatmap(
  gtl = gtl_macrophage,
  geneset_id = res_enrich$gs_id[1],
  FDR = FDR)

GeneTonic::signature_volcano(
  gtl = gtl_macrophage,
  geneset_id = res_enrich$gs_id[1],
  FDR = FDR
)

score_mat <- GeneTonic::gs_scores(gtl = gtl_macrophage)

GeneTonic::gs_scoresheat(score_mat)

GeneTonic::gs_mds(
  gtl = gtl_macrophage,
  n_gs = nrwo(gtl_macrophage$res_enrich),
  mds_labels = 10)

```



# Session information {-}

```{r}
sessionInfo()
```

# Bibliography {-}
